### DISCUSSION TITLE
`lsp-completion-at-point` with `lsp-enable-snippet` causes `corfu-quit` to not cancel preview

### DISCUSSION THREAD

**Discussion by LemonBreezes** [minad/corfu#436] [Q&A]
Hello Minad.

Currently, if you use LSP with `lsp-enable-snippet` and `corfu-preview-current` equal to `insert`, and you preview a candidate which is a snippet for a function then do `corfu-quit`, the candidate preview is not deleted. To illustrate:

Here I have typed out some code:
![image](https://github.com/minad/corfu/assets/29102529/9e7b8d48-5f0d-4d52-b88f-aba8f2b201e3)
Then I hit TAB to preview it:
![image](https://github.com/minad/corfu/assets/29102529/8a7f8d9b-3758-4dcc-8bfc-7b4678375fa3)
Now I hit `C-g` for `corfu-quit` and it gets inserted:
![image](https://github.com/minad/corfu/assets/29102529/f3131ce5-bdae-4437-89b0-423637f95b24)

This might be a bug in `lsp-completion-at-point` itself but I haven't figured it out.

**Comment by minad** [comment]
Hi @LemonBreezes! Unfortunately I don't have much insight in `lsp-completion-at-point` since I haven't used the `lsp-mode` package in a long time. If `corfu--teardown` is executed correctly, then the preview overlay will be deleted:

https://github.com/minad/corfu/blob/46c5ec84e096b852229b32f054462f0a63abfeb5/corfu.el#L1165

I think what you see here is an overlay created by the snippet system and not the Corfu preview. At the point where the snippet system takes over, Corfu is already torn down completely. The `:exit-function` is executed after `corfu-quit`.

**Reply by LemonBreezes** [reply]
Okay, I just looked into this more and realized it was somehow because of Evil that once we hit `<escape>` to quit the completion, it expands the snippet in the case of `lsp-completion-at-point` but not for `yasnippet-capf`. It's bit jarring to leave a snippet active after leaving insert state. I looked into it because of @LuigiPiucco's comment here https://github.com/doomemacs/doomemacs/pull/7002#issuecomment-2008009856 but this seems pretty tricky and I would want to use `lsp-completion-at-point` a lot more before I decide on a change, if any.

**Reply by minad** [reply]
Okay, this seems highly Doom-specific and I guess we can close this here.
